_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,290 @@
+
+In this section, we will also present a proof of correctness for this algorithm based on elementary properties of [probability density function](https://en.wikipedia.org/wiki/Probability_density_function) (shortened as PDF from now on), [cumulative distribution function](https://en.wikipedia.org/wiki/Cumulative_distribution_function) (shortened as CDF from now on), and basic calculus.
+
+First of all, to make sense of all the $\ln(u_j) / w_j$ calculations in this algorithm, one has to understand [inverse transform sampling](https://en.wikipedia.org/wiki/Inverse_transform_sampling). For each $j \in \{1, \dotsc, N\}$, consider the probability distribution defined on $(-\infty, 0)$ with CDF $F_j(x) = e^{w_j \cdot x}$. In order to pluck out a value $y$ from this distribution, we first sample a value $u_j$ uniformly randomly out of $(0, 1)$ that determines the percentile of $y$ (i.e., how our $y$ value ranks relative to all possible $y$ values, a.k.a, the "overall population", from this distribution), and then apply $F_j^{-1}$ to $u_j$ to find $y$, so, $y = F_j^{-1}(u_j) = \ln(u_j) / w_j$.
+
+Secondly, after defining all the required CDF functions $F_j(x) = e^{w_j \cdot x}$ for $j \in \{1, \dotsc, N\}$, we can also easily derive their corresponding PDF functions $f_j$: $$f_j(x) = \frac{d F_j(x)}{dx} = w_j e^{w_j \cdot x}$$.
+
+Finally, with a clear understanding of the family of probability distributions involved, one can prove the probability of this algorithm selecting a given sequence of rows $(r_1, \dotsc, r_n)$ is equal to $\prod\limits_{j = 1}^{n} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$, identical to the probability previously mentioned in the ["What exactly is `SWoR`"](#swor) section, which implies the possible outcomes of this algorithm will follow exactly the same probability distribution as that of a $n$-step `SWoR`.
+
+In order to not deprive our dear readers the pleasure of completing this proof by themselves, we decided to not inline the rest of the proof (which boils down to a calculus exercise) within this blog post, and make everything available in [this file](proof.pdf) instead.
+
+# Weighted sampling with replacement
+
+While all previous sections were solely focused on weighted sampling without replacement, this section will briefly discuss how the exponential-variate approach can also benefit the weighted-sampling-with-replacement use case (which will be shortened as `SWR` from now on).
+
+Although `SWR` with sample size $n$ can obviously be carried out by $n$ independent processes each selecting $1$ sample, parallelizing a `SWR` workload across all partitions of a Spark dataframe (let's call it $X$) will still be more performant if the number of partitions is much larger than $n$ and more than $n$ executors are available in a Spark cluster.
```

Comment:
```suggestion
Although `SWR` with sample size $n$ can be carried out by $n$ independent processes each selecting $1$ sample, parallelizing a `SWR` workload across all partitions of a Spark dataframe (let's call it $X$) will still be more performant if the number of partitions is much larger than $n$ and more than $n$ executors are available in a Spark cluster.
```

Reply:
I just try to remove words like "obviously" where I can, since there's always a chance it's non-obvious to a reader, and just makes them feel bad.